crates/interledger-settlement/src/api.rs

```diff
@@ -157,32 +169,45 @@ impl_web! {
             }})
             .and_then(move |accounts| {
                 let account = &accounts[0];
-                if let Some(settlement_engine) = account.settlement_engine_details() {
-                    Ok((account.clone(), settlement_engine))
+                if account.settlement_engine_details().is_some() {
+                    Ok(account.clone())
                 } else {
                     let error_msg = format!("Account {} does not have settlement engine details configured. Cannot handle incoming settlement", account.id());
                     error!("{}", error_msg);
                     Err((StatusCode::from_u16(404).unwrap(), error_msg))
                 }
             })
-            .and_then(move |(account, settlement_engine)| {
-                let account_id = account.id();
-                let amount: u64 = body.amount.parse().unwrap();
-                let amount = amount.normalize_scale(ConvertDetails {
-                    // We have account, engine and request asset scale.
-                    // Which one is not required? Still not clear from all the spec discussions.
-                    from: account.asset_scale(),
-                    to: settlement_engine.asset_scale
-                });
-                store.update_balance_for_incoming_settlement(account_id, amount, idempotency_key)
+            .and_then(move |account| {
+                result(BigUint::from_str(&amount))
                 .map_err(move |_| {
-                    let error_msg = format!("Error updating balance of account: {} for incoming settlement of amount: {}", account_id, amount);
+                    let error_msg = format!("Could not convert amount: {:?}", amount);
                     error!("{}", error_msg);
                     (StatusCode::from_u16(500).unwrap(), error_msg)
                 })
-            }).and_then(move |_| {
-                let ret = Bytes::from("Success");
-                Ok((StatusCode::OK, ret))
+                .and_then(move |amount_from_engine| {
+                    let account_id = account.id();
+                    let amount = amount_from_engine.normalize_scale(ConvertDetails {
+                        from: engine_scale,
+                        to: account.asset_scale(),
+                    });
+                    // If we'd overflow, settle for the maximum u64 value
+                    let safe_amount = if let Some(amount) = amount.to_u64() {
```

Comment:
I think it's slightly less of a foot-gun to do this instead of using a different variable name:

```suggestion
                    let amount = if let Some(amount) = amount.to_u64() {
```

Reply:
I'd rather be explicit about the variable name here and leave it as `safe_amount`. Why do you think it's a footgun?

Reply:
I like using shadowing when doing these kind of transformations on numbers, just to make sure that I don't accidentally use the wrong one in something after applying the transformation (and to make sure that anyone working on that code later will avoid the same mistake).

Reply:
OK I agree with you.